tracer/src/Datadog.Trace.SourceGenerators/TelemetryMetric/Sources.cs

```diff
@@ -1160,196 +1073,311 @@ private static string GetValues(
-    {
-        sb.AppendLine(
+        WriteTagArgList(sb, tagArray);
+
+        sb.Append(
             $$"""
-                  public void Record{{details.ShortName}}{{property}}({{tagName}} tag, double value)
+              double value)
                   {
-                      var index = {{index}} + (int)tag;
-                      _buffer.{{details.ShortName}}[index].TryEnqueue(value);
-                  }
+                      var index = {{index}}
               """);
-    }
 
-    private static void WriteRecordDistribution(StringBuilder sb, in TelemetryMetricGenerator.EnumDetails details, string property, int index, string tagName1, string tagName2, int tag2EntryCount)
-    {
+        WriteTagIndices(sb, enumDictionary, tagArray);
+
         sb.AppendLine(
             $$"""
-                  public void Record{{details.ShortName}}{{property}}({{tagName1}} tag1, {{tagName2}} tag2, double value)
-                  {
-                      var index = {{index}} + ((int)tag1 * {{tag2EntryCount}}) + (int)tag2;
+              ;
                       _buffer.{{details.ShortName}}[index].TryEnqueue(value);
                   }
               """);
     }
 
-    private static void WriteNoopCount(StringBuilder sb, in TelemetryMetricGenerator.EnumDetails details, string property)
+    private static void WriteTagArgList(StringBuilder sb, string[] tagArray)
     {
-        sb.AppendLine(
-            $$"""
-                  public void Record{{details.ShortName}}{{property}}(int increment = 1)
-                  {
-                  }
-              """);
-    }
+        for (var i = 0; i < tagArray.Length; i++)
+        {
+            sb.Append(tagArray[i])
+              .Append(" tag");
 
-    private static void WriteNoopCount(StringBuilder sb, in TelemetryMetricGenerator.EnumDetails details, string property, string tagName)
-    {
-        sb.AppendLine(
-            $$"""
-                  public void Record{{details.ShortName}}{{property}}({{tagName}} tag, int increment = 1)
-                  {
-                  }
-              """);
+            if (tagArray.Length > 1)
+            {
+                sb.Append(i + 1);
+            }
+
+            sb.Append(", ");
+        }
     }
 
-    private static void WriteNoopCount(StringBuilder sb,  in TelemetryMetricGenerator.EnumDetails details, string property, string tagName1, string tagName2)
+    private static void WriteTagIndices(StringBuilder sb, Dictionary<string, EquatableArray<string>> enumDictionary, string[] tagArray)
     {
-        sb.AppendLine(
-            $$"""
-                  public void Record{{details.ShortName}}{{property}}({{tagName1}} tag1, {{tagName2}} tag2, int increment = 1)
-                  {
-                  }
-              """);
+        if (tagArray.Length == 0)
+        {
+            return;
+        }
+
+        Span<int> multipliers = stackalloc int[tagArray.Length];
```

Comment:
what's the expected length here? be careful with int size * length in the stack...

Reply:
It's one per generic parameter, so currently a max of 5. i.e. all fine IMO